core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+            switch (operation) {
+                case DEFAULT_CASE_OPERATION_NUMBER:
+                    break;
+                case INSERT_OPERATION_NUMBER:
+                    insertPositions++;
+                    break;
+                case DELETE_OPERATION_NUMBER:
+                    deletePositions++;
+                    break;
+                case UPDATE_OPERATION_NUMBER:
+                    updatePositions++;
+                    break;
+                default:
+                    throw new IllegalArgumentException("Unknown operator number " + operation);
+            }
+        }
+
+        int totalPositions = insertPositions + deletePositions + 2 * updatePositions;
+        Block rowIdBlock = inputPage.getBlock(redistributionColumnCount);
+
+        ImmutableList.Builder<Type> pageTypesBuilder = ImmutableList.builder();
+        pageTypesBuilder.addAll(dataColumnTypes);
+        pageTypesBuilder.add(INTEGER);
+        List<Type> pageTypes = pageTypesBuilder.build();
+        PageBuilder pageBuilder = new PageBuilder(totalPositions, pageTypes);
+        int resultPosition = 0;
+        int rowIdPosition = 0;
+        int[] rowIdPositions = new int[totalPositions];
+        for (int position = 0; position < originalPositionCount; position++) {
+            long operation = INTEGER.getLong(operationChannelBlock, position);
+            if (operation != DEFAULT_CASE_OPERATION_NUMBER) {
+                if (operation == DELETE_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addDeleteRow(pageBuilder, inputPage, position);
+                    rowIdPositions[resultPosition] = rowIdPosition;
+                    resultPosition++;
+                }
+                if (operation == INSERT_OPERATION_NUMBER || operation == UPDATE_OPERATION_NUMBER) {
+                    addInsertRow(pageBuilder, mergeCaseBlocks, position);
+                    rowIdPositions[resultPosition] = 0;
+                    resultPosition++;
+                }
+            }
+            if (INTEGER.getLong(rowMatchedBlock, position) > 0) {
+                rowIdPosition++;
+            }
+        }
+        checkArgument(resultPosition == totalPositions, "resultPosition (%s) is not equal to totalPositions (%s)", resultPosition, totalPositions);
+        Page pageWithoutRowIdBlock = pageBuilder.build();
+        int pageChannels = pageWithoutRowIdBlock.getChannelCount();
+        Block[] blocksWithRowId = new Block[pageChannels + 1];
+        for (int channel = 0; channel < pageChannels; channel++) {
+            blocksWithRowId[channel] = pageWithoutRowIdBlock.getBlock(channel);
+        }
+        blocksWithRowId[pageChannels] = makeOutputRowIdBlock(rowIdBlock, totalPositions, rowIdPositions);
+
+        return new Page(blocksWithRowId);
+    }
+
+    private static Block makeOutputRowIdBlock(Block rowIdBlock, int totalPositions, int[] rowIdPositions)
```

Comment:
`totalPositions == rowIdPositions.length`, so `totalPositions` arg is looks redundant

Reply:
Removed.